### DISCUSSION TITLE
[Bug] Updating extra lib not working after dispose in 0.34.0 and 0.39.0 version. It is working fine in lower version 0.26.0

### DISCUSSION THREAD

**Discussion by DharmendraSendhav** [microsoft/monaco-editor#4020] [Q&A]
### Reproducible in vscode.dev or in VS Code Desktop?

- [X] Not reproducible in [vscode.dev](https://vscode.dev) or VS Code Desktop

### Reproducible in the monaco editor playground?

- [ ] Not reproducible in [the monaco editor playground](https://microsoft.github.io/monaco-editor/playground.html)

### Monaco Editor Playground Link

We have requirement where we add library at runtime and dispose the lib. But it is not working properly, I checked in the latest version and lower version but in lower version 0.26.0 it is working but in latest versions 0.39.0 it is not working.

Steps to reproduce the issue.
1) Add extra lib in monaco editor using below method.
var library = "declare class Test {   description:string; } ";
 let testDisposable =  this.monaco.languages.typescript.typescriptDefaults.addExtraLib(library, test.d.ts);

2) Dispose the lib
     testDisposable.dispose();

3) Change the variable Name of description to description123 and add again using extra lib function.

var library = "declare class Test {   description123:string; } ";
 let testDisposable =  this.monaco.languages.typescript.typescriptDefaults.addExtraLib(library, test.d.ts);

first time it will dispose the lib and add the new changed lib but second time it will stop the disposing previous library. and will not show the intelligence for changed Test class

### Monaco Editor Playground Code

```typescript
var library = "declare class Test { description:string; } ";
let testDisposable = this.monaco.languages.typescript.typescriptDefaults.addExtraLib(library, test.d.ts);


testDisposable.dispose();

var library = "declare class Test { description123:string; } ";
let testDisposable = this.monaco.languages.typescript.typescriptDefaults.addExtraLib(library, test.d.ts);
```

### Reproduction Steps

Steps to reproduce the issue.
1) Add extra lib in monaco editor using below method.
var library = "declare class Test {   description:string; } ";
 let testDisposable =  this.monaco.languages.typescript.typescriptDefaults.addExtraLib(library, test.d.ts);

2) Dispose the lib
     testDisposable.dispose();

3) Change the variable Name of description to description123 and add again using extra lib function.

var library = "declare class Test {   description123:string; } ";
 let testDisposable =  this.monaco.languages.typescript.typescriptDefaults.addExtraLib(library, test.d.ts);

first time it will dispose the lib and add the new changed lib but second time it will stop the disposing previous library. and will not show the intelligence for changed Test class

### Actual (Problematic) Behavior

If  we dispose the library and add again with changes, first time it  will work but from the second time it is not working.

### Expected Behavior

it should dispose the previous library and add changed library and also should show the intelligence for change library even if we repeat the steps multiple times.

### Additional Context

We have requirement where we add library at runtime and dispose the lib. But it is not working properly, I checked in the latest version and lower version but in lower version 0.26.0 it is working but in latest versions 0.39.0 it is not working.

Steps to reproduce the issue.
1) Add extra lib in monaco editor using below method.
var library = "declare class Test {   description:string; } ";
 let testDisposable =  this.monaco.languages.typescript.typescriptDefaults.addExtraLib(library, test.d.ts);

2) Dispose the lib
     testDisposable.dispose();

3) Change the variable Name of description to description123 and add again using extra lib function.

var library = "declare class Test {   description123:string; } ";
 let testDisposable =  this.monaco.languages.typescript.typescriptDefaults.addExtraLib(library, test.d.ts);

first time it will dispose the lib and add the new changed lib but second time it will stop the disposing previous library. and will not show the intelligence for changed Test class

**Comment by spahnke** [comment]
Are you using JavaScript as the model language? We have the use case where we add libraries on editor open and remove them again on editor close, depending on which editor is opened, and it works fine with 0.39.0. However, in the case of `d.ts` files you have to register them on the language default that corresponds to the model/editor language not the language of the `d.ts` file.

**Edit**: I clarified what is actually necessary and adjusted the playground link correspondingly.

```js
var library = "declare class Test { description:string; } ";
// if editor language is JavaScript register the d.ts file on the javascriptDefaults object
let testDisposable2 = monaco.languages.typescript.javascriptDefaults.addExtraLib(library, monaco.Uri.file("test.d.ts").toString());
```

[Playground link](https://microsoft.github.io/monaco-editor/playground.html?source=v0.39.0#XQAAAAIzBAAAAAAAAABBqQkHQ5NjdMjwa-jY7SIQ9S7DNlzs5W-mwj0fe1ZCDRFc9ws9XQE0SJE1jc2VKxhaLFIw9vEWSxW3yscw5cR1JVciAKctgAboa7AaZZDHLgrLqPRv_fRv2RyIokydlTdTYccsu6v6jDrc8jQ8ulrCVzjWBVWZVl-YrARXWjHX0-tquUNesdygYRbQM1JUnAt3Sx3dTjbY5JYWauONIVXrKyfhXpbQmonnj54EEd_Gg_0InT9zafYLmQsMth9da9JWIymY4mlxHccv7MQdXmhzEUl350-II4z7KQxkamLRaSUORFZ3PCpLMK9oQR3fL-zT1819LiTZFqzxKuU0xfwHrBydX_6qMREVsbSjPSZD6Wo5plTfX6FzWZnfdTPg4GliqHfM9Ql3x7qS7SAKO1CrlpWTfH7w7581y96-Jx8aWpzBBWBshcutZA1v5KzHEhEqTEPeiuZGckadSXF3RUUExkNE20BC8W9XgwdjTPkt7lVkEJxHR8r_Dam_CbG3yQnedrigeGvaNWG8Q2UG_sn9cKkdVu6a9w9mHvD53YcSSLPWz_Enoz2xqzHhALyjuvSkBtQX-eol6UyzJSimHVi9-IKDvm9rwyIB0yy_ZcUB4v4f5kQYtd30BkpKV7O6aT__8ByLpQ)

**Comment by DharmendraSendhav** [comment]
No, we are using typeScript as model language. I checked with javascript and it is working fine, Issue is only with typescript.

**Comment by hediet** [comment]
Please create a new issue with link to a playground example, so we can reproduce. Also, please use the bisect feature in the playground to figure out in which version it broke. Thanks!

**Reply by DharmendraSendhav** [reply]
I have raised below new issue with Playground links and Reproduction Steps.

https://github.com/microsoft/monaco-editor/issues/4023
